pkg/token/serviceprincipaltoken.go

```diff
@@ -69,91 +73,177 @@ func (p *servicePrincipalToken) Token() (adal.Token, error) {
 
 func (p *servicePrincipalToken) TokenWithOptions(options *azcore.ClientOptions) (adal.Token, error) {
 	emptyToken := adal.Token{}
-	var spnAccessToken azcore.AccessToken
+	var accessToken string
+	var expirationTimeUnix int64
+	var err error
+	scopes := []string{p.resourceID + "/.default"}
 
 	// Request a new Azure token provider for service principal
 	if p.clientSecret != "" {
-		clientOptions := &azidentity.ClientSecretCredentialOptions{
-			ClientOptions: azcore.ClientOptions{
-				Cloud: p.cloud,
-			},
-		}
-		if options != nil {
-			clientOptions.ClientOptions = *options
-		}
-		cred, err := azidentity.NewClientSecretCredential(
-			p.tenantID,
-			p.clientID,
-			p.clientSecret,
-			clientOptions,
-		)
-		if err != nil {
-			return emptyToken, fmt.Errorf("unable to create credential. Received: %w", err)
-		}
-
-		// Use the token provider to get a new token
-		spnAccessToken, err = cred.GetToken(context.Background(), policy.TokenRequestOptions{Scopes: []string{p.resourceID + "/.default"}})
+		accessToken, expirationTimeUnix, err = p.getTokenWithClientSecret(context.Background(), scopes, options)
```

Comment:
same comment about the `context.Background()`

Reply:
Updated to reuse ctx